drf_haystack/serializers.py

```diff
@@ -131,24 +138,69 @@ def get_fields(self):
                     warnings.warn("Field '{field}' already exists in the field list. This *will* "
                                   "overwrite existing field '{field}'".format(field=field_name))
                 field_mapping[field_name] = declared_fields[field_name]
-
         return field_mapping
 
     def to_representation(self, instance):
         """
+        If we have a serializer mapping, use that.  Otherwise, use standard serializer behavior
+        """
+        serializers = getattr(self.Meta, "serializers", None)
+        if not serializers:
+            return self.default_to_representation(instance)
+
+        index = instance.searchindex
+        serializer_class = serializers.get(type(index), None)
+        if not serializer_class:
+            raise ImproperlyConfigured("Could not find serializer for %s in mapping" % index)
+        return serializer_class(context=self._context).to_representation(instance)
+
+    def default_to_representation(self, instance):
+        """
         Since we might be dealing with multiple indexes, some fields might
         not be valid for all results. Do not render the fields which don't belong
         to the search result.
         """
         ret = super(HaystackSerializer, self).to_representation(instance)
+        prefix_field_names = len(getattr(self.Meta, "index_classes")) > 1
+        current_index = self._get_index_class_name(type(instance.searchindex))
         for field in self.fields.keys():
-            if field not in chain(instance.searchindex.fields.keys(), self._declared_fields.keys()):
-                del ret[field]
+            orig_field = field
+            if prefix_field_names:
+                parts = field.split("__")
+                if len(parts) > 1:
+                    index = parts[0][1:]  # trim the preceding '_'
+                    field = parts[1]
+                    if index == current_index:
+                        ret[field] = ret[orig_field]
+                    del ret[orig_field]
+            elif field not in chain(instance.searchindex.fields.keys(), self._declared_fields.keys()):
+                del ret[orig_field]
 
         if hasattr(instance, "highlighted") and instance.highlighted:
             ret["highlighted"] = instance.highlighted[0]
         return ret
 
+    def _get_index_class_name(self, index_cls):
+        """
+        Converts in index model class to a name suitable for use as a field name prefix. A user
+        may optionally specify custom aliases via an 'index_aliases' attribute on the Meta class
+        """
+        cls_name = str(index_cls).replace("<class '", "").replace("'>", "")
```

Comment:
For me it seems a bit safer to use `cls_name = index_cls.__class__.__name__` in order to get the class name. The method you have used is based on the class `__repr__` (or `__str__` can't remember), and feels a bit unsafe to me.